File: Data_Structure/recursion.py
# Factorial
def recursion(n):
    if n == 0:
        return 1
    else:
        return n * recursion(n-1)


# Fibonacci
def fibonacci(n):
    if n == 1:
        return 0
    if n == 2:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)


# Binary Search Tree
def binary_s_recur(arr, first, last, target):
    if first > last:
        return -1
    mid = int((first+last)/2)

    if arr[mid] == target:
        return mid
    elif target < arr[mid]:
        return binary_s_recur(arr, first, mid-1, target)
    else:
        return binary_s_recur(arr, mid+1, last, target)


# Tower of Hanoi
def hanoi_tower(height, from_pole, to_pole, with_pole):
    if height >= 1:
        hanoi_tower(height-1, from_pole, with_pole, to_pole)
        move_disk(from_pole, to_pole)
        hanoi_tower(height-1, with_pole, to_pole, from_pole)

def move_disk(fp, tp):
    print(f"Moving disk from {fp} to {tp}")

File: Data_Structure/recursion.c
// Recursion
#include<stdio.h>

// Factorial
long factorial(int n)
{
    if (n == 0)
        return 1;
    else
        return (n*factorial(n-1));
}

// Binary Search
int BinarySearch(int ar[], int first, int last, int target)
{
    int mid;
    if(first > last)
        return -1;
    mid = (first + last) / 2;

    if(ar[mid] == target)
        return mid;
    else if(target < ar[mid])
        return BinarySearch(ar, first, mid-1, target);
    else
        return BinarySearch(ar, mid+1, last, target); 
}

int main()
{
    // Test Factorial
    int num;
    long fact;

    printf("Enter a Positive Integer\n");
    scanf("%d", &num);

    if (num < 0)
        printf("!! Factorials aren't defined for the Negative Integer !!");
    else
    {
        fact = factorial(num);
        printf("%d! =  %ld\n", num, fact);
    }

    return 0;
}

